sharedingress-config-generator/controller.go

```diff
@@ -0,0 +1,173 @@
+
+	ConfigPath               string
+	HAProxyRuntimeSocketPath string
+
+	// lastReloadedHash stores the hash of the last successfully reloaded configuration
+	lastReloadedHash []byte
+}
+
+func (r *SharedIngressConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.Client = mgr.GetClient()
+
+	// A channel is used to generate an initial sync event.
+	// Afterwards, the controller syncs on HostedClusters.
+	initialSync := make(chan event.GenericEvent)
+
+	go func() {
+		initialSync <- event.GenericEvent{Object: &hyperv1.HostedCluster{}}
+	}()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&hyperv1.HostedCluster{}).
+		WatchesRawSource(source.Channel(initialSync, &handler.EnqueueRequestForObject{})).
+		Watches(
+			&routev1.Route{},
+			handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []ctrl.Request {
+				if _, hasHCPLabel := obj.GetLabels()[util.HCPRouteLabel]; !hasHCPLabel {
+					return nil
+				}
+				return []ctrl.Request{{NamespacedName: client.ObjectKey{
+					Name:      obj.GetName(),
+					Namespace: obj.GetNamespace(),
+				}}}
+			}),
+		).
+		Named("SharedIngressConfigGenerator").
+		Complete(r)
+}
+
+const (
+	reloadTimeout = 5 * time.Second
+)
+
+func (r *SharedIngressConfigReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	currentHash, err := r.currentConfigHash()
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Create a temporary file for the new config
+	dir := filepath.Dir(r.ConfigPath)
+	tmpFile, err := os.CreateTemp(dir, ".haproxy.*.cfg.tmp")
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create temporary config file: %w", err)
+	}
+	defer tmpFile.Close()
+	defer os.Remove(tmpFile.Name())
+
+	// Use MultiWriter to write to both hash and file simultaneously
+	h := sha256.New()
+	writer := io.MultiWriter(tmpFile, h)
+
+	if err := generateRouterConfig(ctx, r.Client, writer); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to generate config: %w", err)
+	}
+	newHash := h.Sum(nil)
+
+	// Update the config file if it doesn't exist or has changed
+	if currentHash == nil || !bytes.Equal(currentHash, newHash) {
+		logger.Info("HAProxy configuration change detected!")
+
+		// Ensure the file is synced to disk before we rename it
+		if err := tmpFile.Sync(); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to sync temporary file: %w", err)
+		}
+		tmpFile.Close()
+
+		if err := os.Rename(tmpFile.Name(), r.ConfigPath); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update config file: %w", err)
+		}
+
+		logger.Info("HAProxy configuration file updated", "hash", fmt.Sprintf("%x", newHash))
+	}
+
+	// Reload HAProxy if current config differs from last successfully reloaded config
+	if !bytes.Equal(r.lastReloadedHash, newHash) {
+		logger.Info("Reloading HAProxy configuration")
+		if response, err := sendHAProxyCommand(r.HAProxyRuntimeSocketPath, "reload"); err != nil {
```

Comment:
What are possible responses, any docs we can link? is it possible we don't get an error but still a response we don't want?

Reply:
> isn't this missing a
> global
> master-worker
> 
> stanza in the config template https://www.haproxy.com/documentation/haproxy-configuration-tutorials/proxying-essentials/configuration-basics/programs/

This is done when running the HAProxy container with the flags `-W -S <socket>`